compiler/modules/front/expr.bal

```diff
@@ -972,18 +972,18 @@ function selectMappingInherentType(ExprContext cx, t:SemType expectedType, s:Map
         return cx.semanticErr("ambiguous inherent type for mapping constructor", s:range(expr));
     }
     t:SemType semType = alts[0].semType;
-    mat = t:mappingAtomicType(tc, semType);
+    mat = t:mappingAtomicDerefType(tc, semType);
     if mat is () {
         return cx.semanticErr("applicable type for mapping constructor is not atomic", s:range(expr));
     }
     return [semType, mat];
 }
 
-function mappingAlternativeAllowsFields(t:MappingAlternative alt, string[] fieldNames) returns boolean {
+function mappingAlternativeAllowsFields(t:Context cx, t:MappingAlternative alt, string[] fieldNames) returns boolean {
     t:MappingAtomicType? pos = alt.pos;
     if pos !is () {
         // SUBSET won't be right with record defaults
-        if pos.rest == t:NEVER {
+        if t:simpleCellAtomicType(cx, pos.rest).t == t:NEVER {
```

Comment:
I think we should introduce a function in the types module that takes a mapping atomic type and returns the type inside the cell of the rest field